orchestrator/src/compression/stateful.rs

```
@@ -0,0 +1,237 @@
+        let mut keys: HashSet<Felt> = HashSet::new();
+
+        // Collecting all the keys for which mapping might be required
+        state_update.state_diff.storage_diffs.iter().for_each(|diff| {
+            // Skip the special address
+            if ValueMapping::skip(diff.address) {
+                return;
+            }
+            keys.insert(diff.address);
+            diff.storage_entries.iter().for_each(|entry| {
+                keys.insert(entry.key);
+            });
+        });
+        state_update.state_diff.deployed_contracts.iter().for_each(|contract| {
+            keys.insert(contract.address);
+        });
+        state_update.state_diff.nonces.iter().for_each(|nonce| {
+            keys.insert(nonce.contract_address);
+        });
+        state_update.state_diff.replaced_classes.iter().for_each(|replaced_class| {
+            keys.insert(replaced_class.contract_address);
+        });
+
+        // Fetch the values for the keys from the special address (0x2)
+        let special_address_mappings = ValueMapping::get_special_address_mappings(state_update)?;
+
+        // Fetch the value for the keys in special address either from the current mapping or from the provider
+        // Doing this in parallel
+        let fetch_results: Vec<_> = stream::iter(keys)
+            .map(|key| {
+                let special_address_mappings = special_address_mappings.clone();
+                async move {
+                    match special_address_mappings.get(&key).cloned() {
+                        Some(value) => Ok((key, value)),
+                        None => {
+                            Ok((key, ValueMapping::get_value_from_provider(provider, &key, pre_range_block).await?))
+                        }
+                    }
+                }
+            })
+            .buffer_unordered(3000)
+            .collect()
+            .await;
+
+        for (key, value) in parse_results(fetch_results)? {
```

Isn’t there a way to fail fast on a stream iterator?

---

yep, you can use [`try_fold`](https://docs.rs/futures/latest/futures/stream/trait.TryStreamExt.html#method.try_fold) from `TryStreamExt`

---

Or just [`try_collect`](https://docs.rs/futures/latest/futures/stream/trait.TryStreamExt.html#method.try_collect) for that matter.

---

Updated!